### PR TITLE
Internal: Set image projects via configs

### PR DESCRIPTION
**Pull request by Subbarker**
## Description

This is a clean start from https://github.com/GoogleCloudPlatform/ops-agent/pull/1643

Right now we interpret image names passed in from configs to determine what project they come from, and where they are an "Image" or an "Image Family". This logic has grown in complexity over time and most importantly, requires changes to our testing logic whenever a new corner case shows up.

This PR implements storing project names alongside the image/family names with test configs in a way that is simple to manage for future tests. We are calling these strings "Image Specs"

E.g.:

Image Family / Project: `<project>:<family>`

Specific Image / Project: `<project>=<image>`

## Related issue

b/330177642

## Checklist:
- Unit tests
  - [x] Unit tests do not apply.
  - [ ] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [ ] Integration tests do not apply.
  - [x] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

File: integration_test/gce/gce_testing.go [hunk 1]
```diff
@@ -1151,12 +1114,21 @@ func attemptCreateInstance(ctx context.Context, logger *log.Logger, options VMOp
 	}
 
 	imgProject := options.ImageProject
+
```
Comment (martijnvans) [inline integration_test/gce/gce_testing.go:1117]: I wish I could put this comment on line 1850, but github won't let me do that.

This PR is changing the meaning of VMOptions.Platform, VMOptions.Image, and VMOptions.ImageProject and the comments should be updated accordingly. [sidenote: the "Required" comment on VMOptions.Platform is incorrect. It is required that VMOptions.Platform OR VMOptions.Image are specified though.] Also callsites that use these fields should be updated. Now I'm not sure exactly what you want the new fields to look like and how they should relate to each other. Should everybody cram everything into options.Platform, which then parses it back out? Or should callsites that want to deal with the new format of the PLATFORMS env variable do the parsing, populating the appropriate fields of VMOptions like they do today? (except VMOptions.ImageProject would be required now).

Basically, VMOptions is the interface to this function and I want the interface to be self-consistent and documented.

File: integration_test/gce/gce_testing.go [hunk 2]
```diff
@@ -1151,12 +1114,21 @@ func attemptCreateInstance(ctx context.Context, logger *log.Logger, options VMOp
...
+	delim := ""
+	if strings.Contains(options.Platform, ":"){
+		delim = ":"
+	} else if strings.Contains(options.Platform, "="){
+		delim = "="
+	} else {
+		return nil, fmt.Errorf("could not parse options.Platform from struct: %+v", options)
+	}
+ 
+	s := strings.Split(options.Platform, delim)
```
Comment (martijnvans) [inline integration_test/gce/gce_testing.go:1127]: There are comments, e.g. the one at the top of this file, about what the PLATFORMS env variable should look like and those are all wrong after this PR. They should be updated.

File: project.yaml [hunk 3]
```diff
@@ -6,31 +6,32 @@ targets:
...
   bullseye:
     package_extension:
       deb
     architectures:
       x86_64:
         test_distros:
           representative:
-          - debian-11
+          - debian-cloud:debian-11
+          - ml-images:common-gpu-debian-11-py310
```
Comment (martijnvans) [inline project.yaml:22]: this line should be in a follow up PR.

Comment (Subbarker) [reply project.yaml:22]: Done

File: integration_test/gce/gce_testing.go [hunk 4]
```diff
@@ -1167,15 +1136,27 @@ func attemptCreateInstance(ctx context.Context, logger *log.Logger, options VMOp
 		return nil, fmt.Errorf("attemptCreateInstance() could not construct valid labels: %v", err)
 	}
 
-	imageOrImageFamilyFlag := "--image=" + options.Image
 
-	if options.Platform != "" {
-		imageOrImageFamilyFlag = "--image-family=" + options.Platform
 
-		if image, ok := overriddenImageFamilies[options.Platform]; ok {
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1175]: Done

File: project.yaml [hunk 5]
```diff
@@ -116,56 +117,56 @@ targets:
...
+          - suse-cloud:sles-15
           exhaustive:
-          - sles-15-sp2-sap
-          - sles-15-sp5-sap
-          - opensuse-leap
-          - opensuse-leap-15-4
+          - suse-sap-cloud:sles-15-sp2-sap
+          - suse-sap-cloud:sles-15-sp5-sap
+          - opensuse-cloud:opensuse-leap
+          - opensuse-cloud=opensuse-leap-15-4-v20231208-x86-64
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply project.yaml:151]: Why? this seems correct to me.

[the remark this answers is not in the document]
Comment (martijnvans) [reply project.yaml:151]: why is this PR changing this line though? https://screenshot.googleplex.com/5Ri285yfB7HuUKj

[the remark this answers is not in the document]
Comment (Subbarker) [reply project.yaml:151]: because that's the image getting tested, it's overridden here: https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/gce/gce_testing.go#L1028

File: integration_test/gce/gce_testing.go [hunk 6]
```diff
@@ -1167,15 +1136,27 @@ func attemptCreateInstance(ctx context.Context, logger *log.Logger, options VMOp
...
-			imageOrImageFamilyFlag = "--image=" + image
-		}
+	delim := ""
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go]: Done

File: integration_test/gce/gce_testing.go [hunk 7]
```diff
@@ -1151,12 +1114,21 @@ func attemptCreateInstance(ctx context.Context, logger *log.Logger, options VMOp
 	}
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1144]: Spoken offline, and agree that the long term solution is to replace instances of `platform` with `imageSpec` and updated logic. A followup PR will do a sweeping refactor along these lines.

File: integration_test/gce/gce_testing.go (hunk 1 above)
Comment (Subbarker) [reply integration_test/gce/gce_testing.go]: c+p from another comment: "Spoken offline, and agree that the long term solution is to replace instances of platform with imageSpec and updated logic. A followup PR will do a sweeping refactor along these lines."

File: integration_test/gce/gce_testing.go (hunk 2 above)
Comment (Subbarker) [reply integration_test/gce/gce_testing.go]: c+p from another comment: "Spoken offline, and agree that the long term solution is to replace instances of platform with imageSpec and updated logic. A followup PR will do a sweeping refactor along these lines."

File: integration_test/metadata/integration_metadata.go [hunk 8]
```diff
@@ -132,7 +133,9 @@ func UnmarshalAndValidate(data []byte, i interface{}) error {
 
 func SliceContains(slice []string, toFind string) bool {
 	for _, entry := range slice {
-		if entry == toFind {
+		// vm.Platform (ex. `sles-15`) does not align with ImageSpec (ex. `suse-cloud:sles-15`)
+		// TODO, remove lingering references to "Platform" and use "Image Spec" instead.
+		if strings.HasSuffix(entry, toFind) {
```
Comment (martijnvans) [inline integration_test/metadata/integration_metadata.go:138]: I am confident that SliceContains is not the right place to be working around this. SliceContains is just supposed to be the same as slices.Contains (https://pkg.go.dev/slices#Contains), recently introduced into the language. Looking at the 4 callsites to this function, 2 of them want to look for an exact match, and 2 of them have this problem where they are passing in vm.Platform and comparing that to a list of ImageSpecs. Instead of doing that, i think we should store an ImageSpec in the VM struct alongside the Platform and pass that in to metadata.SliceContains instead of vm.Platform. Longer term, we slowly migrate everything to refer to vm.ImageSpec and get rid of vm.Platform.

File: integration_test/gce/gce_testing.go [hunk 9]
```diff
@@ -1103,11 +1060,50 @@ func getVMPlatform(image string, platform string) (string, error) {
 	return "", errors.New("at least one of image or platform must be specified")
 }
 
+func parseImageSpec(options *VMOptions) (error) {
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1063]: Done

File: integration_test/gce/gce_testing.go [hunk 10]
```diff
@@ -1874,6 +1863,9 @@ func SetupLogger(t *testing.T) *logging.DirectoryLogger {
 
 // VMOptions specifies settings when creating a VM via CreateInstance() or SetupVM().
 type VMOptions struct {
+	// Optional. Can be used to pass image/image family & image project in one
+	// string. If set, Platform/Image/ImageProject should not be set.
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1867]: Done

File: integration_test/soak_test/cmd/launcher/main.go [hunk 11]
```diff
@@ -122,11 +122,11 @@ func mainErr() error {
 
 	// Create the VM.
 	options := gce.VMOptions{
-		Platform:    distro,
-		TimeToLive:  ttl,
-		Name:        vmName,
-		MachineType: "e2-standard-16",
-		Metadata: map[string]string{
+		ImageSpec:    distro,
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/soak_test/cmd/launcher/main.go:125]: Done

File: integration_test/gce/gce_testing.go [hunk 12]
```diff
@@ -1333,6 +1317,11 @@ func CreateInstance(origCtx context.Context, logger *log.Logger, options VMOptio
 	ctx, cancel := context.WithTimeout(origCtx, 3*vmInitTimeout)
 	defer cancel()
 
+	err := parseImageSpec(&options)
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1320]: Done

File: integration_test/metadata/integration_metadata.go (hunk 8 above)
Comment (Subbarker) [reply integration_test/metadata/integration_metadata.go:138]: Done

File: integration_test/third_party_apps_test.go [hunk 13]
```diff
@@ -867,7 +867,7 @@ var gpuModels = map[string]accelerator{
 }
 
 const (
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/third_party_apps_test.go:869]: Done

File: integration_test/gce/gce_testing.go [hunk 14]
```diff
@@ -295,6 +295,7 @@ type VM struct {
 	Project     string
 	Network     string
 	Platform    string
+	ImageSpec   string
```
Comment (martijnvans) [inline integration_test/gce/gce_testing.go:298]: this should be documented too. So should `Platform` but that's outside the scope of this PR. Everything else is self-explanatory (or already has a comment).

File: integration_test/gce/gce_testing.go [hunk 15]
```diff
@@ -1102,22 +1060,60 @@ func getVMPlatform(image string, platform string) (string, error) {
 
 	return "", errors.New("at least one of image or platform must be specified")
 }
```
Comment (martijnvans) [inline integration_test/gce/gce_testing.go:1062]: put a blank line between functions please.

File: integration_test/third_party_apps_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (martijnvans) [reply integration_test/third_party_apps_test.go:869]: Three more: 

https://github.com/GoogleCloudPlatform/ops-agent/blob/8a04b6a7c630a4fc687ebdebf4364ae1aa78970e/integration_test/README.md?plain=1#L53

https://github.com/GoogleCloudPlatform/ops-agent/blob/8a04b6a7c630a4fc687ebdebf4364ae1aa78970e/tasks.mak#L96

https://github.com/GoogleCloudPlatform/ops-agent/blob/8a04b6a7c630a4fc687ebdebf4364ae1aa78970e/dev-docs/new-distro.md?plain=1#L102

File: integration_test/gce/gce_testing.go (hunk 15 above)
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1062]: Fixed.

File: integration_test/gce/gce_testing.go (hunk 14 above)
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:298]: Fixed

File: integration_test/gce/gce_testing.go [hunk 16]
```diff
@@ -1102,22 +1060,60 @@ func getVMPlatform(image string, platform string) (string, error) {
...
 	}
 	vm := &VM{
-		Project:  options.Project,
-		Platform: platform,
-		Name:     options.Name,
-		Network:  os.Getenv("NETWORK_NAME"),
-		Zone:     options.Zone,
+		Project:   options.Project,
+		Platform:  platform,
+		ImageSpec: options.ImageSpec,
```
[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/gce/gce_testing.go:1128]: Done

File: integration_test/third_party_apps_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (martijnvans) [reply integration_test/third_party_apps_test.go:869]: still needs addressing

[the remark this answers is not in the document]
Comment (Subbarker) [reply integration_test/third_party_apps_test.go:869]: Done

File: dev-docs/new-distro.md [hunk 17]
```diff
@@ -100,7 +100,7 @@ The instructions are very similar to the instructions for `ops_agent_test`.
         or
         [test/third_party_apps/presubmit/bullseye_aarch64.gcl](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/kokoro/config/test/third_party_apps/bullseye_aarch64.gcl),
         set `platforms` to `['$DISTRO_FAMILY']` (instead of
-        `['debian-11']`). You should also add this section right after the
+        `['debian-cloud:debian-11']`). You should also add this section right after the
```
Comment (martijnvans) [inline dev-docs/new-distro.md:103]: there are other places in this file that need updating, like line 49 and lines 8-11.

**Review by martijnvans — APPROVED** [review]
